library/std/src/sys/unix/fs.rs

```diff
@@ -1648,55 +1667,155 @@ mod remove_dir_impl {
         }
     }
 
-    fn remove_dir_all_recursive(parent_fd: Option<RawFd>, p: &Path) -> io::Result<()> {
-        let pcstr = cstr(p)?;
-
-        // entry is expected to be a directory, open as such
-        let fd = openat_nofollow_dironly(parent_fd, &pcstr)?;
-
-        // open the directory passing ownership of the fd
-        let (dir, fd) = fdreaddir(fd)?;
-        for child in dir {
-            let child = child?;
-            match is_dir(&child) {
-                Some(true) => {
-                    remove_dir_all_recursive(Some(fd), Path::new(&child.file_name()))?;
-                }
-                Some(false) => {
-                    cvt(unsafe { unlinkat(fd, child.name_cstr().as_ptr(), 0) })?;
-                }
-                None => match cvt(unsafe { unlinkat(fd, child.name_cstr().as_ptr(), 0) }) {
+    fn unlink_direntry(ent: &DirEntry, parent_fd: RawFd) -> io::Result<bool> {
+        match is_dir(&ent) {
+            Some(true) => Ok(false),
+            Some(false) => {
+                cvt(unsafe { unlinkat(parent_fd, ent.name_cstr().as_ptr(), 0) })?;
+                Ok(true)
+            }
+            None => {
+                match cvt(unsafe { unlinkat(parent_fd, ent.name_cstr().as_ptr(), 0) }) {
                     // type unknown - try to unlink
                     Err(err)
                         if err.raw_os_error() == Some(libc::EISDIR)
                             || err.raw_os_error() == Some(libc::EPERM) =>
                     {
-                        // if the file is a directory unlink fails with EISDIR on Linux and EPERM everyhwere else
-                        remove_dir_all_recursive(Some(fd), Path::new(&child.file_name()))?;
+                        // if the file is a directory unlink fails with EISDIR on Linux
+                        // and EPERM everyhwere else
+                        Ok(false)
                     }
-                    result => {
-                        result?;
-                    }
-                },
+                    result => result.map(|_| true),
+                }
             }
         }
+    }
 
-        // unlink the directory after removing its contents
-        cvt(unsafe {
-            unlinkat(parent_fd.unwrap_or(libc::AT_FDCWD), pcstr.as_ptr(), libc::AT_REMOVEDIR)
+    struct CachedReadDir {
+        readdir: ReadDir,
+        raw_fd: RawFd,
```

Comment:
TODO: Get rid of this and use `dirfd()`instead.

Reply:
This should be fine as long as the field isn't accessed directly. You can add a function that returns a `BorrowedFd` with the the lifetime matching self borrow. That avoids a few extra calls into libc so we don't have to get the descriptor out of the DIR struct every time we need it.

We should avoid using `RawFd` when things can be given a lifetime.

Reply:
In particular since I just found this tidbit in the dirfd(3) man page: 
>       ENOTSUP
>              The implementation does not support the association of a file descriptor with a di‐
>              rectory.

Reply:
Hmm, that appears to be a Linux-ism. [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/functions/dirfd.html) does not mention failure with ENOTSUP.